spec/connection.md

```diff
@@ -0,0 +1,258 @@
+# NATS Connection
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2023-10-12 |
+|Author  |@Jarema |
+|Status  |`Implemented`|
+|Tags    |client, server|
+
+|Revision|Date|Author|Info|
+ |--------|----|------|----|
+ |1       |2023-10-12|@author|Initial draft|
+
+## Summary
+
+This document describes how clients connect to the NATS server or NATS cluster.
+That includes topics like:
+- connection process
+- reconnect
+- tls
+- discoverability of other nodes in a cluster
+
+## Motivation
+
+Ensuring a consistent way how Clients establish and maintain connection with the NATS server and provide consistent and predictable behaviour across the ecosystem.
+
+## Guide-level Explanation
+
+### Establishing the connection
+
+**TODO** Add WebSocket flow.
+
+#### Minimal example
+
+1. Clients initiate a network connection to the Server.
+2. Server responds with [INFO][INFO] json.
+3. Client sends [CONNECT][CONNECT] json.
+4. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+**Note** Server can send subsequent [INFO][INFO] on a ongoing connection.
```

Comment:
Server will do so only if the client sends `1` or above for the `protocol` field in the `[CONNECT][CONNECT]` protocol.